docs/contributing/999-rfcs/2024-03-14-explicit-lift-qualification.md

```diff
@@ -0,0 +1,246 @@
+
+In the future we'd like to employ data flow analysis (probably through [reaching defintions analysis](https://en.wikipedia.org/wiki/Reaching_definition)) resolve these errors.
+
+## Problems with the current solution
+
+We need a mechanism for explicitly qualifying lifted objects. Consider the following cases:
+
+1. When using inflight expressions to reference a preflight object we'd like to disable the compilation error and explicitly qualify the lift.
+```js
+let b1 = new cloud.Bucket() as "b1";
+let b2 = new cloud.Bucket() as "b2";
+inflight () => {
+  let var b = b1;
+  if true {
+    b = b2;
+  }
+  b.put("key", "value"); // Need to explicitly qualify b1 and b2 with "put" since the compiler doesn't figure this out for us.
+}
+```
+
+2. When there's a collection of preflight objects that are being used inflight:
+```js
+for b in buckets {
+  b.put("key", "value"); // We need to qualify all `buckets` with "put"
+}
+```
+3. When our knowledge of the application logic can produce better qualifications:
+```js
+for i in buckets.len {
+  if i % 2 == 0 {
+    buckets.at(i).put("key", "value"); // We only need to qualify the even buckets with "put" and not the odd ones
+  }
+}
+```
+4. Cases where just the method name isn't enough information to create least privilege permissions:
+```js
+for i in 0..5 {
+  bucket.put("k{i}", "value");
+}
+```
+Here we can benefit from an explicit qualification of `bucket` with `"put"` operation and a specific key pattern: `"k*"`. We currently don't
+have a way to disable the automatic qualification and define the narrower qualification.
+
+Note that in all the above exmaples except the last one we currently get a compilation error due to us trying to use a `Bucket.put` operation on an inflight expression.
+
+## Requirements
+
+So the aim of this proposal is to present:
+1. A way to explicitly qualify the preflight objects that are being used by inflight methods/closures.
+2. Disabling compilation errors where the explicit qualification applies.
+3. Disabling automatic qualification where the explicit qualification applies. 
+4. Provide a framework that can be extended to programatically defining and reusing lift qualifications.
+
+## Naming
+
+"qualifiations" is used because it qualifies how the preflight object is being used. 
+Other words that we sometime used in the past and might be clearer are "capabilities" or "operations":
+* "capabilities" makes sense because it describes the capabilities the host needs in order to successfully run the inflight code.
+* "operations" makes sense because it describes the operations being performed on the preflight objects by the inflight code.
+
+In this proposal I'll use "qualifications", but I'd be happy to hear what you think.
```

Comment:
Since this is a parameter of `lift`, I think we can't use the term without pairing it with the term "lift".

"Lift x"

`std.LiftQualification` is such a mouthful.

Other ideas:

- "lift specification", and then we can use `std.LiftSpec` 
- "lift definition" (`std.LiftDef`).
-  `LiftScope` (might be too overloaded with node scope). 
- `LiftContext` 👍


My favorite is `LiftContext`. It also feels like it will be a good vessel to add additional attributes to the lift in the future.